pkg/ingester/instance_limits.go

```diff
@@ -46,12 +47,38 @@ func (l *InstanceLimits) RegisterFlags(f *flag.FlagSet) {
 }
 
 // Sets default limit values for unmarshalling.
-var defaultInstanceLimits *InstanceLimits
+var (
+	defaultInstanceLimits    *InstanceLimits
+	defaultInstanceLimitsMtx sync.RWMutex
```

Comment:
We should call `setDefaultInstanceLimits` from `initRuntimeConfig`, before we start using this default value: https://github.com/grafana/mimir/blob/d3c05df8fd02a64e9d1b47a6f37add6a94640e47/pkg/mimir/modules.go#L270-L271

We don't need a mutex here, because all uses of default value happen after setting the default.

Reply:
The value is set when a new ingester instances is created. We triggered the race detector with Distributor unit tests but it seems like we don't have any ingester ones that worked the same way. In either case, I really don't like mutable global state so I'm going to keep the mutex (it's RW so reads should be very cheap).

Reply:
I don't agree with this decision. I think introducing unnecessary lock here is confusing. Reader may rightly ask why is it there, if it's not needed. I still suggest we remove it.

Reply:
Changed